### ISSUE TITLE
The Tale of the Github Repository

### ISSUE THREAD

**Issue by MichaelPaulukonis** [dariusk/NaNoGenMo-2015#162]
Github repositories are full of events. Can we frame these events as a story?

OF COURSE WE CAN!!!

Can we make this story _interesting_?

Let's avoid answering that, shall we?

Code and notes are here: https://github.com/MichaelPaulukonis/NaNoGenMo2015/tree/master/github-narrative

I have a small proof-of-concept gist [here](https://gist.github.com/MichaelPaulukonis/663a224d7faf40a698d6).

Extracts:

> On Sun Oct 25 2015 19:27:52, dariusk opened a new issue called "Resources". But it's an admin issue, so who cares?
> On Sun Oct 25 2015 19:28:38, dariusk opened a new issue called "Procedural Visual Novel".
> On Sun Oct 25 2015 19:55:08, bcj opened a new issue called "An Attempt to Exhaust Memory to Simulate a Place to Attempt to Exhaust".
> On Sun Oct 25 2015 22:41:47, zachwhalen opened a new issue called "Using a password dump for a corpus".
> On Wed Oct 28 2015 16:18:43, tra38 opened a new issue called "The Atheists Who Believe In God". And it's been completed. Sweet!

The plan is to dig deeper -- getting time-frames for when labels were placed (so we know what was completed first, etc), interesting events (opened a 2nd issue, completed a second project, etc). If the word count is absolutely abominable -- I also have the entire body of comments to play with.

My original thought was to tell the "story" of my past year on GitHub -- but only the last 300 events/90-days of events are available via the API, so I switched tracks.

**Comment by MichaelPaulukonis** [comment]
It has not escaped my attention that this is also the story of NaNoGenMo-2015.

**Comment by MichaelPaulukonis** [comment]
@tra38 - you might find this interesting. You and @hugovk were some of the (in)direct inspirations for this.

**Comment by tra38** [comment]
Yeah, I do find this interesting. I once had the idea of using information from GitHub to write a narrative, but I couldn't think of a good way of doing it. Will be monitoring progress.

**Comment by MichaelPaulukonis** [comment]
It needs work, but it has nearly 100K words: https://gist.github.com/MichaelPaulukonis/1b95c25f7eca4942933a

Of course, that's because it's including ALL THE COMMENTS. ugh.

A future elaboration will have a summary of the comments, if they are > a certain number of characters.

Another elaboration is to try to suss out the target repo, and extract the languages used. We'll see how this goes.

Of some interest may be the ranking algorithm, which ranks issues. It's more of a proof-of-concept, but a start.

```
if (iss.comments) { atom.rank += iss.comments.length; }

// rank by labels
if (iss.labelTypes.preview) { atom.rank += 5; }
if (iss.labelTypes.completed) { atom.rank += 20; }
if (iss.labelTypes.closed) { atom.rank += -1; } // hrm...
// admin issues aren't as interesting....
if (!iss.labelTypes.admin) { atom.rank += 5; }
```

a better comment-ranking algorithm would give more weight to non-author comments, for example.
Also, cross-references should have some weight, time-to-completion, how early opened, notable firsts, etc.

As of 2015.11.30, 10:39pm EST:

The top 20 issues, ranked:

_Compiler pipeline + writers' techniques = a "proper novel" ::blink::_ was opened by cpressey.

_Simulationist Fantasy Novel_ was opened by mattfister.

_"Where I'm From" poem & novel generator_ was opened by marythought.

_A daring journey to the bottom of the pit_ was opened by mcwill97.

_It takes a "Village" to translate "Hamlet"_ was opened by dkurth.

_The Atheists Who Believe In God_ was opened by tra38.

_Co-authored Procedural Novel_ was opened by dariusk.

_Cheating pseudo-entry: Vocabulary mashup_ was opened by mewo2.

_Generative Socratic Dialogues_ was opened by yourpalal.

_The Null Earth Catalog_ was opened by coleww.

_intense intents in tents_ was opened by emdaniels.

_Browne Garden Commonplace Book_ was opened by spenteco.

_Interlude: (Un)Sound Structures_ was opened by jseakle.

_The TPP: A "Found" Generated Novel_ was opened by coleww.

_Saga III: Another Original Play by a Computer_ was opened by lizadaly.

_Goal-driven use of scenes and sequels for capers_ was opened by enkiv2.

_Molly's Feed_ was opened by moonmilk.

_(NaNoGenMo: Dadism 2.0) 2.0_ was opened by tra38.

_A play based on the "french 4chan"_ was opened by WhiteFangs.

_Neuralgae_ was opened by spikelynch.

**Comment by MichaelPaulukonis** [comment]
@hugovk - it hits 50K, stick a fork in it and call it _done_.

**Comment by hugovk** [comment]
@MichaelPaulukonis Splat.

I like that's it's also pulled in the images.

> Another elaboration is to try to suss out the target repo, and extract the languages used.

https://developer.github.com/v3/repos/#list-languages

**Comment by MichaelPaulukonis** [comment]
Thank you, sir!

Yup, but each issue has the repo linked in a different way -- so it'll be a combination of parsing, plus a manual whitelist for those that never indicated it, or that link to multiple repos (say, when just commenting on somebody else's project).

But THEN I'll be able to complete the Language survey for each year!

...finally.
